src/clean-core/allocator.hh

```diff
@@ -142,55 +165,126 @@ struct stack_allocator final : allocator
     void reset()
     {
         _head = _buffer_begin;
-        _last_alloc = nullptr;
+        _last_alloc_id = 0;
     }
 
     stack_allocator() = default;
-    stack_allocator(span<std::byte> buffer) : _buffer_begin(buffer.data()), _head(buffer.data()), _buffer_end(buffer.data() + buffer.size()) {}
+    explicit stack_allocator(span<std::byte> buffer) : _buffer_begin(buffer.data()), _head(buffer.data()), _buffer_end(buffer.data() + buffer.size())
+    {
+    }
 
 private:
     std::byte* _buffer_begin = nullptr;
     std::byte* _head = nullptr;
     std::byte* _buffer_end = nullptr;
-    std::byte* _last_alloc = nullptr;
+    int32_t _last_alloc_id = 0;
 };
 
-
-/// for short lived allocations serviced from a ring buffer
-/// can optionally fall back to an external allocator if overcommited
-struct scratch_allocator final : allocator
+/// linear allocator operating in virtual memory
+/// reserves pages on init, commits pages on demand
+/// only frees pages if explicitly called
+struct virtual_linear_allocator final : allocator
```

Comment:
Hm we now have a lot of "niche" allocators in this general allocator header that is included everywhere. I think we should all/most specialized allocators into a separate file, probably even experimental for now

Reply:
Yes, in fact an allocators/ folder might be a good idea with separate headers per type. We could still have "sink headers" for multiple ones